fix: Remove `no-extra-parens` autofix for potential directives

fasttime opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

References #16988 

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

This PR fixes an issue with the `no-extra-parens` rule that could cause the autofix to change the behavior of the code. The solution consists in disabling the autofix in situations where it would otherwise introduce a new directive.
This solution has been discussed in issue #16988, and I started working on the `no-extra-parens` because I think that there is sufficient consensus on how to handle the autofix in the case of this specific rule.

The problem addressed here is that a statement like `("use strict");` can mutate into a `use strict` directive if the parentheses are removed, and that would change the semantics of the code. To determine whether the autofix should be disabled, we decided to check if a string literal is the sole child of an `ExpressionStatement`.
Since there is no danger that the `no-extra-parens` and `quotes` rule fix the same node concurrently (see https://github.com/eslint/eslint/pull/17022#discussion_r1154150126), it is safe for `no-extra-parens` to remove parentheses from a template literal.

#### Is there anything you'd like reviewers to focus on?

fasttime commented on lib/rules/utils/ast-utils.js:
```diff
@@ -1501,6 +1501,37 @@ module.exports = {
+     * @returns {boolean} Whether a specified node is an ExpressionStatement that is either part of,
+     * or immediately follows a (possibly empty) directive prologue.
+     */
+    isExpressionInOrJustAfterDirectivePrologue(sourceCode, node) {
```
This is a refactoring of the deleted function `isPartOfDirectivePrologue` from `quotes.js` with few changes:
* Renamed the function to `isExpressionInOrJustAfterDirectivePrologue` to better describe what it does
* Changed the argument to the be the parent node (the `ExpressionStatement`)
* Added a check to make sure that the specified node is really an `ExpressionStatement`
* Fixed to logic that checks if a sibling node is a directive to ensure that literals in parentheses test negative
* Added an additional `SourceCode` argument to the function (necessary to test for parentheses)

nzakas commented on lib/rules/no-extra-parens.js:
```diff
@@ -386,6 +386,24 @@ module.exports = {
+         * @private
+         */
+        function isFixable(node) {
+            if (!(node.type === "Literal" && typeof node.value === "string")) {
```
Can we refactor this a bit to make it easier to understand? And maybe add a comment?

```suggestion
            // if it's not a string literal it can be autofixed
            if (node.type !== "Literal" || typeof node.value !== "string")) {
```

fasttime commented:
> Can also update the documentation to note this change?

Thanks @nzakas, I've added a sentence to the Rule Details. I think there is no special way to format a note like this?

nzakas commented on lib/rules/no-extra-parens.js:
```diff
@@ -395,7 +395,8 @@ module.exports = {
          */
         function isFixable(node) {
-            if (!(node.type === "Literal" && typeof node.value === "string")) {
+            // if it's not a string literal it can be autofixed
```
```suggestion

            // if it's not a string literal it can be autofixed
```

mdjermanovic commented on lib/rules/no-extra-parens.js:
```diff
@@ -386,6 +386,31 @@ module.exports = {
+            if (isParenthesisedTwice(node)) {
+                return true;
+            }
+            return !astUtils.isExpressionInOrJustAfterDirectivePrologue(sourceCode, node.parent);
```
I thought we'd just check if the parent is ExpressionStatement as we don't know what fixes other rules are applying in the code that precedes it.

For example:

```js
/* eslint no-extra-parens: "error" */
/* eslint prettier/prettier: "error" */

; //
("use strict");

```

after `--fix`, `"use strict"` still ends up being a directive:

```js
/* eslint no-extra-parens: "error" */
/* eslint prettier/prettier: "error" */

//
"use strict";

```

fasttime replied:
Well, if compatibility with Prettier is a concern then we could just check the parent node as you suggest. I don't know how we could stop third-party rules from interfering if another autofix is applied to the `ExpressionStatement` node itself or to its children rather than to a sibling node (e.g. `("use " + "strict");` → `("use strict");`), but from what I can tell, that is not what Prettier does.

In this case we should also change the behavior of the `quotes` rule, because fox example

```js
/* eslint quotes: "error" */
/* eslint prettier/prettier: "error" */

; //
`use strict`;
```

autofixes to

```js
/* eslint quotes: "error" */
/* eslint prettier/prettier: "error" */

//
"use strict";
```

mdjermanovic replied:
> Well, if compatibility with Prettier is a concern then we could just check the parent node as you suggest.

Not specifically with Prettier, any custom or third-party rules.

> I don't know how we could stop third-party rules from interfering if another autofix is applied to the `ExpressionStatement` node itself or to its children rather than to a sibling node (e.g. `("use " + "strict");` → `("use strict");`),

That shouldn't cause problems because the ranges will be overlapping so ESLint would apply only one of the fixes:

https://eslint.org/docs/latest/extend/custom-rules#conflicting-fixes

I think that non-directive ExpressionStatement > Literal is a very edge case, and it's a meaningless code that should be manually fixed anyway, so I'd vote for the safest solution which is also the simplest for us. @nzakas what do you think?

> ```js
> /* eslint quotes: "error" */
> /* eslint prettier/prettier: "error" */
> 
> ; //
> `use strict`;
> ```

Yes, I don't think this should be autofixed either.

fasttime replied:
> That shouldn't cause problems because the ranges will be overlapping so ESLint would apply only one of the fixes:
> 
> https://eslint.org/docs/latest/extend/custom-rules#conflicting-fixes

Ah, understood. I didn't realize that removing the parentheses is handled as one single fix by the linter. Thanks for the explanation @mdjermanovic.

nzakas commented on docs/src/rules/no-extra-parens.md:
```diff
@@ -21,6 +21,8 @@ This rule always ignores extra parentheses around the following:
 * immediately-invoked function expressions (also known as IIFEs) such as `var x = (function () {})();` and `var x = (function () {}());` to avoid conflicts with the [wrap-iife](wrap-iife) rule
 * arrow function arguments to avoid conflicts with the [arrow-parens](arrow-parens) rule
 
+Problems reported by this rule can be fixed automatically, except when removing the parentheses would create a new directive.
```
I was thinking more that we need to show some examples, as I don't know that most people will understand what "removing the parentheses would create a new directive" actually means.

fasttime replied:
Thanks for the feedback @nzakas! Would it be also fine for you if we disable the autofix for all string literals that are direct children of an `ExpressionStatement`, as suggested [in this comment](https://github.com/eslint/eslint/pull/17022#discussion_r1152293435) by @mdjermanovic? This would improve interoperability with third-party rules at the expenses of a few autofixes in some rare edge cases.

Rec0iL99 commented:
Don't close. Not stale.

> Would it be also fine for you if we disable the autofix for all string literals that are direct children of an ExpressionStatement, as suggested https://github.com/eslint/eslint/pull/17022#discussion_r1152293435 by @mdjermanovic?

@nzakas there's a reply from @fasttime for you to look into here. Thanks.

nzakas commented:
Thanks, replied!

fasttime commented on lib/rules/quotes.js:
```diff
@@ -328,12 +330,11 @@ module.exports = {
                     },
                     fix(fixer) {
-                        if (isPartOfDirectivePrologue(node)) {
+                        if (node.parent.type === "ExpressionStatement" && !astUtils.isParenthesised(sourceCode, node)) {
```
Shall we also check if the `ExpressionStatement` is at the top level of a program or function body?

fasttime commented on lib/rules/no-extra-parens.js:
```diff
@@ -386,6 +386,30 @@ module.exports = {
+            if (isParenthesisedTwice(node)) {
+                return true;
+            }
+            return node.parent.type !== "ExpressionStatement";
```
Shall we also check if the `ExpressionStatement` is at the top level of a program or function body?

mdjermanovic replied:
I think yes, if it isn't at the top level of a program or function body then the fix seems safe.

fasttime replied:
> I think yes, if it isn't at the top level of a program or function body then the fix seems safe.

Thanks! This is done in https://github.com/eslint/eslint/pull/17022/commits/7a2800d663b5fe5633c2f807c5a856796e6c2496. I also took the time to refactor similar code in other rules that do the same kind of checks.

fasttime commented on docs/src/rules/no-extra-parens.md:
````diff
@@ -21,6 +21,25 @@ This rule always ignores extra parentheses around the following:
+```
+
+In this case, the rule will not try to remove the parentheses around `"use strict"` but will still report them as a problem.
+
````
Do we want to add this information to `quotes.md` as well?

fasttime commented:
Thanks for the suggestions @mdjermanovic @nzakas. I've updated the autofix logic as per discussion above and I've added an example to the rule docs that shows the effect of the disabled autofix. I've also reverted my changes in `ast-utils` and moved `isExpressionInOrJustAfterDirectivePrologue()` back into `quotes.js`, but still keeping the new name and the fix that checks for parentheses. Let me know if I overlooked something.

mdjermanovic commented:
There's a merge conflict.

fasttime commented:
Rebased to fix the merge conflict.

mdjermanovic commented on tests/lib/rules/utils/ast-utils.js:
```diff
@@ -1802,4 +1802,62 @@ describe("ast-utils", () => {
+            linter.defineRule("checker", {
+                create: mustCall(context => ({
+                    ":expression": mustCall(node => {
+                        const sourceCode = context.getSourceCode();
+
+                        assert.strictEqual(astUtils.isTopLevelExpressionStatement(sourceCode, node), false);
```
```suggestion
                        assert.strictEqual(astUtils.isTopLevelExpressionStatement(node), false);
```

`astUtils.isTopLevelExpressionStatement` takes only 1 argument.

mdjermanovic commented on tests/lib/rules/utils/ast-utils.js:
```diff
@@ -1802,4 +1802,62 @@ describe("ast-utils", () => {
+            it(`should return ${expectedRetVal} for \`${nodeText}\` in \`${code}\``, () => {
+                linter.defineRule("checker", {
+                    create: mustCall(context => {
+                        const sourceCode = context.getSourceCode();
```
```suggestion
                        const sourceCode = context.sourceCode;
```

`context.getSourceCode()` is now deprecated (https://github.com/eslint/eslint/pull/17107).

mdjermanovic commented on tests/lib/rules/no-extra-parens.js:
```diff
@@ -3444,6 +3444,22 @@ ruleTester.run("no-extra-parens", rule, {
-        )
+        ),
+
+        // Potential directives (no autofix)
+        invalid("('use strict');", null),
+        invalid("function f() { ('abc'); }", null),
+        invalid("(function () { ('abc'); })();", null),
+        invalid("_ = () => { ('abc'); };", null),
+        invalid("'use strict';(\"foobar\");", null),
```
I think these all would really become directives. It would be good to also add a test where the expression statement wouldn't end up in the directive prologue so that we confirm that the no-autofix behavior in such case is intentional.

For example: `invalid("foo(); ('bar');", null),`

fasttime commented:
Thanks for the suggestions @mdjermanovic, it's all updated in https://github.com/eslint/eslint/commit/aa16e4c5b642074d85470136c471f01eb2409c5e now.

fasttime commented on lib/rules/quotes.js:
```diff
@@ -212,7 +213,7 @@ module.exports = {
                 // Directive Prologues.
                 case "ExpressionStatement":
-                    return isPartOfDirectivePrologue(node);
+                    return !astUtils.isParenthesised(sourceCode, node) && isExpressionInOrJustAfterDirectivePrologue(node);
```
> However, since the PR is about no-extra-parens, and this change isn't just refactoring, it might be better to extract all changes to the quotes rule into a separate PR?

Agreed. These fixes in the `quotes` rule were not expected from my side, they came more as a side effect of refactoring. I'm going to revert these changes for now and I will submit a new PR to update `quotes` once the changes in `ast-utils.js` are merged.

mdjermanovic reviewed (approved):
LGTM, thanks! Leaving it open for @nzakas to verify.

mdjermanovic commented:
@fasttime this should also be rebased because it includes a change in the docs.

fasttime commented:
> @fasttime this should also be rebased because it includes a change in the docs.

Rebased on top of main. I have no idea why the EasyCLA check is failing.

snitin315 commented:
/easycla

snitin315 commented:
I think the bug in CLA is when a commit is co-authored. Maybe try amending the commit to remove the co-author?

snitin315 reviewed (approved):
LGTM, thanks!

fasttime commented:
> I think the bug in CLA is when a commit is co-authored. Maybe try amending the commit to remove the co-author?

Thanks @snitin315, I hadn't noticed that there is a problem with the EasyCLA bot. Now that you mention it, since this is not the only affected PR, maybe we should just wait until the issue is fixed?

nzakas reviewed (approved):
LGTM. Thanks!
